File: apps/api/app/api/v1/content.py
"""Content upload and DNA card API routes — Firestore-backed."""

from datetime import datetime, timezone
from uuid import uuid4

from fastapi import APIRouter, Depends, Query, status

from app.core.firestore import get_db
from app.middleware.auth import get_current_user
from app.utils.exceptions import NotFoundError, ValidationError
from app.utils.storage import generate_upload_url

router = APIRouter()


@router.post("/upload-url")
async def get_upload_url(
    request_body: dict,
    current_user=Depends(get_current_user),
) -> dict:
    """Generate a signed upload URL for Firebase Storage."""
    upload_url, storage_path = generate_upload_url(
        file_name=request_body.get("file_name", ""),
        content_type=request_body.get("content_type", ""),
    )
    return {"upload_url": upload_url, "storage_path": storage_path}


@router.post("/upload", status_code=status.HTTP_201_CREATED)
async def upload_content(
    request_body: dict,
    current_user=Depends(get_current_user),
) -> dict:
    """Upload content for analysis and DNA card generation."""
    raw_content = request_body.get("raw_content")
    storage_path = request_body.get("storage_path")

    if not raw_content and not storage_path:
        raise ValidationError(
            message="Content required",
            detail="Either raw_content or storage_path must be provided",
        )

    db = get_db()
    now = datetime.now(timezone.utc)
    doc_id = str(uuid4())

    content_data = {
        "user_id": current_user.id,
        "title": request_body.get("title", ""),
        "content_type": request_body.get("content_type", "text"),
        "raw_content": raw_content,
        "storage_path": storage_path,
        "source_url": request_body.get("source_url"),
        "status": "uploaded",
        "content_dna": None,
        "created_at": now,
        "updated_at": now,
    }
    await db.collection("content_uploads").document(doc_id).set(content_data)

    return {**content_data, "id": doc_id}


@router.get("")
async def list_content(
    page: int = Query(1, ge=1),
    page_size: int = Query(20, ge=1, le=100),
    current_user=Depends(get_current_user),
) -> dict:
    """List the user's content uploads with pagination."""
    db = get_db()

    # Fetch all user content ordered by created_at desc
    query = (
        db.collection("content_uploads")
        .where("user_id", "==", current_user.id)
        .order_by("created_at", direction="DESCENDING")
    )

    all_items = []
    async for doc in query.stream():
        all_items.append({**doc.to_dict(), "id": doc.id})

    total = len(all_items)
    offset = (page - 1) * page_size
    items = all_items[offset : offset + page_size]

    return {"items": items, "total": total}


@router.get("/{content_id}")
async def get_content(
    content_id: str,
    current_user=Depends(get_current_user),
) -> dict:
    """Get a single content upload with its DNA card."""
    db = get_db()
    doc = await db.collection("content_uploads").document(content_id).get()

    if not doc.exists:
        raise NotFoundError(message="Content not found", detail=f"No content upload found with id {content_id}")

    data = doc.to_dict()
    if data.get("user_id") != current_user.id:
        raise NotFoundError(message="Content not found", detail=f"No content upload found with id {content_id}")

    return {**data, "id": doc.id}


@router.post("/{content_id}/analyze")
async def analyze_content(
    content_id: str,
    current_user=Depends(get_current_user),
) -> dict:
    """Trigger or re-trigger AI analysis on existing content."""
    db = get_db()
    doc = await db.collection("content_uploads").document(content_id).get()

    if not doc.exists:
        raise NotFoundError(message="Content not found", detail=f"No content upload found with id {content_id}")

    data = doc.to_dict()
    if data.get("user_id") != current_user.id:
        raise NotFoundError(message="Content not found", detail=f"No content upload found with id {content_id}")

    # Mark as analyzing
    await db.collection("content_uploads").document(content_id).update({"status": "analyzing"})

    # Return current state (analysis will be triggered async via Cloud Tasks)
    data["status"] = "analyzing"
    return {**data, "id": doc.id}


@router.patch("/{content_id}")
async def update_content(
    content_id: str,
    request_body: dict,
    current_user=Depends(get_current_user),
) -> dict:
    """Update emphasis notes or focus settings on a content upload."""
    db = get_db()
    doc = await db.collection("content_uploads").document(content_id).get()

    if not doc.exists:
        raise NotFoundError(message="Content not found", detail=f"No content upload found with id {content_id}")

    data = doc.to_dict()
    if data.get("user_id") != current_user.id:
        raise NotFoundError(message="Content not found", detail=f"No content upload found with id {content_id}")

    # Update the DNA card with emphasis/focus metadata
    dna = data.get("content_dna") or {}
    update_fields = {k: v for k, v in request_body.items() if v is not None}
    if update_fields:
        dna["user_adjustments"] = update_fields
        await db.collection("content_uploads").document(content_id).update({
            "content_dna": dna,
            "updated_at": datetime.now(timezone.utc),
        })
        data["content_dna"] = dna

    return {**data, "id": doc.id}


@router.delete("/{content_id}", status_code=status.HTTP_204_NO_CONTENT)
async def delete_content(
    content_id: str,
    current_user=Depends(get_current_user),
) -> None:
    """Delete a content upload."""
    db = get_db()
    doc = await db.collection("content_uploads").document(content_id).get()

    if not doc.exists:
        raise NotFoundError(message="Content not found", detail=f"No content upload found with id {content_id}")

    data = doc.to_dict()
    if data.get("user_id") != current_user.id:
        raise NotFoundError(message="Content not found", detail=f"No content upload found with id {content_id}")

    await db.collection("content_uploads").document(content_id).delete()

File: functions/src/api/content.ts
/**
 * Content API — 7 onCall functions: getUploadURL, createContent, getContent, listContent, updateContent, triggerGeneration, reanalyzeContent.
 */

import { onCall } from "firebase-functions/v2/https";
import { FieldValue, Timestamp } from "firebase-admin/firestore";
import { v4 as uuidv4 } from "uuid";
import { db, storage } from "../config/firebase.js";
import { ANTHROPIC_API_KEY } from "../config/env.js";
import { verifyAuth } from "../middleware/auth.js";
import { assertRole } from "../middleware/rbac.js";
import { validate } from "../middleware/validate.js";
import { Collections } from "../shared/collections.js";
import { ContentUploadRequestSchema, ContentUpdateRequestSchema, GenerateRequestSchema } from "../shared/schemas.js";
import { wrapError, NotFoundError } from "../shared/errors.js";
import { docToResponse } from "../shared/transform.js";
import { enqueueTask, getTaskHandlerUrl } from "../lib/taskClient.js";

// ─── getUploadURL ────────────────────────────────────────────────────────────
export const getUploadURL = onCall(async (request) => {
  try {
    const ctx = await verifyAuth(request);
    assertRole(ctx, "editor");

    const fileName = `content/${ctx.workspaceId}/${uuidv4()}`;
    const bucket = storage.bucket();
    const file = bucket.file(fileName);

    const [url] = await file.getSignedUrl({
      version: "v4",
      action: "write",
      expires: Date.now() + 15 * 60 * 1000, // 15 minutes
      contentType: "application/octet-stream",
    });

    return { upload_url: url, storage_path: fileName };
  } catch (err) {
    throw wrapError(err);
  }
});

// ─── createContent ───────────────────────────────────────────────────────────
export const createContent = onCall({ secrets: [ANTHROPIC_API_KEY] }, async (request) => {
  try {
    const ctx = await verifyAuth(request);
    assertRole(ctx, "editor");
    const input = validate(ContentUploadRequestSchema, request.data);

    const docRef = db.collection(Collections.CONTENT_UPLOADS).doc();
    const docData = {
      workspaceId: ctx.workspaceId,
      title: input.title,
      contentType: input.content_type,
      rawContent: input.raw_content || "",
      storagePath: input.storage_path || null,
      sourceUrl: input.source_url || null,
      contentDna: null,
      status: "pending",
      createdAt: FieldValue.serverTimestamp(),
      updatedAt: FieldValue.serverTimestamp(),
    };

    await docRef.set(docData);

    // Enqueue content analysis task if content is available
    if (input.raw_content) {
      try {
        await docRef.update({ status: "analyzing" });
        await enqueueTask({
          queue: "content-analysis",
          url: getTaskHandlerUrl("taskContentAnalysis"),
          payload: {
            contentId: docRef.id,
            workspaceId: ctx.workspaceId,
          },
        });
      } catch (taskErr) {
        console.error("Failed to enqueue content analysis:", taskErr);
        // Don't fail the create — the task can be retried manually
      }
    }

    const snap = await docRef.get();
    return docToResponse(docRef.id, snap.data() as Record<string, unknown>);
  } catch (err) {
    throw wrapError(err);
  }
});

// ─── getContent ──────────────────────────────────────────────────────────────
export const getContent = onCall(async (request) => {
  try {
    const ctx = await verifyAuth(request);
    const contentId = request.data?.content_id as string;
    if (!contentId) throw new NotFoundError("content_id required");

    const snap = await db.collection(Collections.CONTENT_UPLOADS).doc(contentId).get();
    if (!snap.exists) throw new NotFoundError("Content not found");

    const data = snap.data() as Record<string, unknown>;
    if (data.workspaceId !== ctx.workspaceId) throw new NotFoundError("Content not found");

    return docToResponse(snap.id, data);
  } catch (err) {
    throw wrapError(err);
  }
});

// ─── listContent ─────────────────────────────────────────────────────────────
export const listContent = onCall(async (request) => {
  try {
    const ctx = await verifyAuth(request);
    const limit = Math.min(request.data?.limit ?? 20, 100);
    const offset = request.data?.offset ?? 0;

    let query = db
      .collection(Collections.CONTENT_UPLOADS)
      .where("workspaceId", "==", ctx.workspaceId)
      .orderBy("createdAt", "desc");

    // For pagination, use offset approach (simpler, suitable for moderate data)
    if (offset > 0) {
      query = query.offset(offset);
    }
    query = query.limit(limit);

    const snap = await query.get();
    const items = snap.docs.map((doc) => docToResponse(doc.id, doc.data() as Record<string, unknown>));

    // Get total count (separate query)
    const countSnap = await db
      .collection(Collections.CONTENT_UPLOADS)
      .where("workspaceId", "==", ctx.workspaceId)
      .count()
      .get();

    return {
      items,
      total: countSnap.data().count,
    };
  } catch (err) {
    throw wrapError(err);
  }
});

// ─── updateContent ───────────────────────────────────────────────────────────
export const updateContent = onCall(async (request) => {
  try {
    const ctx = await verifyAuth(request);
    assertRole(ctx, "editor");
    const contentId = request.data?.content_id as string;
    if (!contentId) throw new NotFoundError("content_id required");
    const input = validate(ContentUpdateRequestSchema, request.data);

    const docRef = db.collection(Collections.CONTENT_UPLOADS).doc(contentId);
    const snap = await docRef.get();
    if (!snap.exists) throw new NotFoundError("Content not found");
    if ((snap.data() as Record<string, unknown>).workspaceId !== ctx.workspaceId) {
      throw new NotFoundError("Content not found");
    }

    // Store user adjustments in the DNA
    const currentDna = (snap.data() as Record<string, unknown>).contentDna as Record<string, unknown> | null;
    const updatedDna = {
      ...(currentDna || {}),
      userAdjustments: {
        emphasisNotes: input.emphasis_notes ?? null,
        focusHookIndex: input.focus_hook_index ?? null,
        additionalContext: input.additional_context ?? null,
      },
    };

    await docRef.update({
      contentDna: updatedDna,
      updatedAt: FieldValue.serverTimestamp(),
    });

    const updated = await docRef.get();
    return docToResponse(docRef.id, updated.data() as Record<string, unknown>);
  } catch (err) {
    throw wrapError(err);
  }
});

// ─── triggerGeneration ───────────────────────────────────────────────────────
export const triggerGeneration = onCall({ secrets: [ANTHROPIC_API_KEY] }, async (request) => {
  try {
    const ctx = await verifyAuth(request);
    assertRole(ctx, "editor");
    const contentId = request.data?.content_id as string;
    if (!contentId) throw new NotFoundError("content_id required");
    const input = validate(GenerateRequestSchema, request.data);

    const docRef = db.collection(Collections.CONTENT_UPLOADS).doc(contentId);
    const snap = await docRef.get();
    if (!snap.exists) throw new NotFoundError("Content not found");

    const data = snap.data() as Record<string, unknown>;
    if (data.workspaceId !== ctx.workspaceId) throw new NotFoundError("Content not found");

    // Update status to generating
    await docRef.update({ status: "generating", updatedAt: FieldValue.serverTimestamp() });

    // Enqueue output generation task
    await enqueueTask({
      queue: "output-generation",
      url: getTaskHandlerUrl("taskOutputGeneration"),
      payload: {
        contentId: docRef.id,
        workspaceId: ctx.workspaceId,
        voiceProfileId: input.voice_profile_id || null,
        selectedPlatforms: input.selected_platforms || null,
        emphasisNotes: input.emphasis_notes || null,
      },
    });

    return { success: true, message: "Generation started. Outputs will appear shortly." };
  } catch (err) {
    throw wrapError(err);
  }
});

// ─── reanalyzeContent ───────────────────────────────────────────────────────
export const reanalyzeContent = onCall({ secrets: [ANTHROPIC_API_KEY] }, async (request) => {
  try {
    const ctx = await verifyAuth(request);
    assertRole(ctx, "editor");
    const contentId = request.data?.content_id as string;
    if (!contentId) throw new NotFoundError("content_id required");

    const docRef = db.collection(Collections.CONTENT_UPLOADS).doc(contentId);
    const snap = await docRef.get();
    if (!snap.exists) throw new NotFoundError("Content not found");

    const data = snap.data() as Record<string, unknown>;
    if (data.workspaceId !== ctx.workspaceId) throw new NotFoundError("Content not found");

    await docRef.update({ status: "analyzing", updatedAt: FieldValue.serverTimestamp() });

    await enqueueTask({
      queue: "content-analysis",
      url: getTaskHandlerUrl("taskContentAnalysis"),
      payload: {
        contentId: docRef.id,
        workspaceId: ctx.workspaceId,
      },
    });

    return { success: true, message: "Content re-analysis started." };
  } catch (err) {
    throw wrapError(err);
  }
});
